src/tcache.c

```diff
@@ -134,6 +134,151 @@ tcache_gc_item_delay_compute(szind_t szind) {
+		nshard_remote += (cache_bin_sz_t)(addr_cur < shard_min_heuristic ||
+		    addr_cur >= shard_max_heuristic);
+	}
+	assert(nslab_remote >= nshard_remote && nslab_remote <= ncached);
+	assert(addr_min != NULL && addr_max != NULL);
+	/*
+	 * Number of ptrs from heuristic remote shards is more than the intended
+	 * flush amount, use the shard addr range as the anchor for flushing.
+	 */
+	if (nshard_remote >= nflush) {
+		*addr_min = shard_min_heuristic;
+		*addr_max = shard_max_heuristic;
+		return nshard_remote;
+	}
+
+	/*
+	 * Not enough ptrs are found from the heuristic shard addr, expand the
+	 * remote addr range further to any addr that's not from the local slab.
+	 */
+	*addr_min = slab_min;
+	*addr_max = slab_max;
+	return nslab_remote;
+}
+
+/* Shuffle the ptrs in the bin to put the remote pointers at the bottom. */
+static inline void
+tcache_gc_small_bin_shuffle(cache_bin_t *cache_bin, cache_bin_sz_t nremote,
+   uintptr_t addr_min, uintptr_t addr_max) {
+	void **swap = NULL;
+	cache_bin_sz_t ncached = cache_bin_ncached_get_local(cache_bin);
+	cache_bin_sz_t ntop = ncached - nremote, cnt = 0;
+	assert(ntop > 0 && ntop < ncached);
+	/*
+	 * Scan the [head, head + ntop) part of the cache bin, during which
+	 * bubbling the non-remote ptrs to the top of the bin.
+	 * After this, the [head, head + cnt) part of the bin contains only
+	 * non-remote ptrs, and they're in the same relative order as before.
+	 * While the [head + cnt, head + ntop) part contains only remote ptrs.
+	 */
+	void **head = cache_bin->stack_head;
+	for (void **cur = head; cur < head + ntop; cur++) {
+		assert(*cur != NULL);
+		if ((uintptr_t)*cur >= addr_min &&
+		    (uintptr_t)*cur < addr_max) {
+			/* Tracks the number of non-remote ptrs seen so far. */
+			cnt++;
+			/*
+			 * There is remote ptr before the current non-remote ptr,
+			 * swap the current non-remote ptr with the remote ptr,
+			 * and increment the swap pointer so that it's still
+			 * pointing to the top remote ptr in the bin.
+			 */
+			if (swap != NULL) {
+				assert(swap < cur);
+				assert((uintptr_t)*swap < addr_min ||
+				    (uintptr_t)*swap >= addr_max);
+				void *tmp = *cur;
+				*cur = *swap;
+				*swap = tmp;
+				swap++;
+				assert(swap <= cur);
+				assert((uintptr_t)*swap < addr_min ||
+				    (uintptr_t)*swap >= addr_max);
+			}
+			continue;
+		} else if (swap == NULL) {
+			/* Swap always points to the top remote ptr in the bin. */
+			swap = cur;
+		}
+	}
+	/*
+	 * Scan the [head + ntop, head + ncached) part of the cache bin,
+	 * after which it should only contain remote ptrs.
+	 */
+	for (void **cur = head + ntop; cur < head + ncached; cur++) {
+		assert(*cur != NULL);
+		/* Early break if all non-remote ptrs have been moved. */
+		if (cnt == ntop) {
+			break;
+		}
+		if ((uintptr_t)*cur >= addr_min && (uintptr_t)*cur < addr_max) {
+			assert(*(head + cnt) != NULL && (
+			    (uintptr_t)*(head + cnt) < addr_min ||
+			    (uintptr_t)*(head + cnt) >= addr_max));
```

Comment:
Judging whether a pointer is remote or local is used multiple times. Maybe put it into a function? This makes it neater.